cgen error for `mut static levels := 0; defer { levels-- }`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 74080e8, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 d32969ed3ea071346c5d110d9c68cc09e604cc64.74080e8
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 20.04.6 LTS
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
|Memory              |0.45GB/15.05GB
|                    |
|V executable        |/home/delian/v/v
|V last modified time|2025-06-23 08:37:09
|                    |
|V home dir          |OK, value: /home/delian/v
|VMODULES            |OK, value: /home/delian/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/delian/v
|                    |
|env VFLAGS          |"-no-parallel"
|Git version         |git version 2.49.0
|V git status        |0.4.11-25-g74080e8f
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|gcc version         |gcc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|clang version       |clang version 10.0.0-4ubuntu1 
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.31-0ubuntu9.16) 2.31

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  a.v && a`
```v
@[unsafe]
fn g() {
	mut static levels := 0
	levels++
	defer { levels-- }
}
unsafe { g() }

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/a.01JYDZV43MAQ9N252X3WRAGYAM.tmp.c:5595: error: redeclaration of 'levels'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

**What did you expect to see?**

a compiled program


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/cmd/tools/vast/vast.v
+++ b/cmd/tools/vast/vast.v
@@ -956,6 +956,8 @@ fn (t Tree) var(node ast.Var) &Node {
 	obj.add_terse('expr', t.expr(node.expr))
 	obj.add_terse('is_arg', t.bool_node(node.is_arg))
 	obj.add_terse('is_mut', t.bool_node(node.is_mut))
+	obj.add_terse('is_static', t.bool_node(node.is_static))
+	obj.add_terse('is_volatile', t.bool_node(node.is_volatile))
 	obj.add('is_used', t.bool_node(node.is_used))
 	obj.add('is_changed', t.bool_node(node.is_changed))
 	obj.add_terse('ct_type_var', t.enum_node(node.ct_type_var))

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -904,6 +904,8 @@ pub:
 	name            string
 	share           ShareType
 	is_mut          bool
+	is_static       bool
+	is_volatile     bool
 	is_autofree_tmp bool
 	is_inherited    bool
 	has_inherited   bool

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -187,11 +187,8 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 			g.write('static ')
 		}
 	}
-	if node.is_volatile {
-		is_defer_var := node.left[0] is ast.Ident && node.left[0].name !in g.defer_vars
-		if !is_defer_var {
-			g.write('volatile ')
-		}
+	if node.is_volatile && node.left[0] is ast.Ident && node.left[0].name !in g.defer_vars {
+		g.write('volatile ')
 	}
 	mut return_type := ast.void_type
 	is_decl := node.op == .decl_assign

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -179,10 +179,22 @@ fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type)
 fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 	mut node := unsafe { node_ }
 	if node.is_static {
-		g.write('static ')
+		is_defer_var := node.left[0] is ast.Ident && node.left[0].name in g.defer_vars
+		if is_defer_var && node.op == .decl_assign {
+			return
+		}
+		if !is_defer_var {
+			g.write('static ')
+		}
 	}
 	if node.is_volatile {
-		g.write('volatile ')
+		is_defer_var := node.left[0] is ast.Ident && node.left[0].name !in g.defer_vars
+		if is_defer_var && node.op == .decl_assign {
+			return
+		}
+		if !is_defer_var {
+			g.write('volatile ')
+		}
 	}
 	mut return_type := ast.void_type
 	is_decl := node.op == .decl_assign

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -189,9 +189,6 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 	}
 	if node.is_volatile {
 		is_defer_var := node.left[0] is ast.Ident && node.left[0].name !in g.defer_vars
-		if is_defer_var && node.op == .decl_assign {
-			return
-		}
 		if !is_defer_var {
 			g.write('volatile ')
 		}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -444,6 +444,12 @@ fn (mut g Gen) gen_fn_decl(node &ast.FnDecl, skip bool) {
 					}
 					info := var.obj as ast.Var
 					if g.table.sym(info.typ).kind != .function {
+						if info.is_static {
+							g.write('static ')
+						}
+						if info.is_volatile {
+							g.write('volatile ')
+						}
 						g.writeln('${g.styp(info.typ)}${deref} ${c_name(var.name)};')
 					}
 				}

--- a/vlib/v/parser/assign.v
+++ b/vlib/v/parser/assign.v
@@ -224,6 +224,8 @@ fn (mut p Parser) partial_assign_stmt(left []ast.Expr) ast.Stmt {
 						expr:         if left.len == right.len { right[i] } else { ast.empty_expr }
 						share:        share
 						is_mut:       lx.is_mut || p.inside_for
+						is_static:    is_static
+						is_volatile:  is_volatile
 						pos:          lx.pos
 						is_stack_obj: p.inside_for
 					}

--- a/vlib/v/tests/defer/defer_static_test.v
+++ b/vlib/v/tests/defer/defer_static_test.v
@@ -1,0 +1,35 @@
+@[unsafe]
+fn g() {
+	mut static levels := 0
+	levels++
+	defer { levels-- }
+}
+
+fn f(depth int) {
+	if depth == 0 {
+		return
+	}
+	unsafe {
+		mut static levels := 0
+		levels++
+		defer { levels-- }
+		if depth == 3 {
+			assert levels == 1
+		}
+		if depth == 2 {
+			assert levels == 2
+		}
+		if depth == 1 {
+			assert levels == 3
+		}
+		println('levels: ${levels} | depth: ${depth}')
+	}
+	f(depth - 1)
+}
+
+fn test_main() {
+	f(3)
+	f(3)
+
+	unsafe { g() }
+}

